clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/PubSubBasedAdminClientManager.java

```diff
@@ -114,74 +103,80 @@ private Future<SubscriptionAdminClient> getOrCreateSubscriptionAdminClient() {
      *
      * @param endpoint The endpoint name of the topic, e.g. command_internal.
      * @param prefix The prefix of the topic, e.g. the adapter instance ID.
-     * @return A succeeded Future if the topic is successfully created or already exists, or a failed
-     *         Future if it could not be created.
+     * @return A succeeded Future if the topic is successfully created or already exists, or a failed Future if it could
+     *         not be created.
      */
     public Future<String> getOrCreateTopic(final String endpoint, final String prefix) {
         final String topicName = PubSubMessageHelper.getTopicName(endpoint, prefix);
         final TopicName topic = TopicName.of(projectId, topicName);
 
-        if (topics.contains(topic.toString())) {
-            LOG.debug("Topic {} already exists, continue", topic);
-            return Future.succeededFuture(topic.getTopic());
-        }
         return getOrCreateTopicAdminClient()
                 .onFailure(thr -> LOG.debug("admin client creation failed", thr))
-                .compose(client -> getOrCreateTopic(projectId, topic, client));
+                .compose(client -> getTopic(topic, client)
+                        .recover(thr -> createTopic(projectId, topic, client)));
+    }
+
+    private Future<String> getTopic(final TopicName topic, final TopicAdminClient client) {
+        try {
+            return Future.succeededFuture(client.getTopic(topic).getName());
+        } catch (ApiException e) {
+            return Future.failedFuture(String.format("Could not get topic. %s", e));
```

Comment:
That is not why I meant by *including the cause*.
I had something in mind like this:
```java
return Future.failedFuture(new ServerErrorException(503, "could not get topic " + topic.getName(), e));
```

Reply:
ok, I updated it